tests/unit/contracts/test_forbidden.py

```diff
@@ -187,6 +187,53 @@ def test_wildcards_in_source_modules_are_resolved(self):
             ],
         }
 
+    def test_recursive_wildcards_in_source_modules_are_resolved(self):
+        graph = self._build_graph()
+        contract = self._build_contract(
+            forbidden_modules=("mypackage.green"),
+            source_modules=("mypackage.one.**",),
+            include_external_packages=False,
+        )
+
+        check = contract.check(graph=graph, verbose=False)
+        assert check.metadata == {
+            "invalid_chains": [
+                {
+                    "upstream_module": "mypackage.green",
+                    "downstream_module": "mypackage.one.alpha",
+                    "chains": [
+                        [
+                            {
+                                "importer": "mypackage.one.alpha",
+                                "imported": "mypackage.green.beta",
+                                "line_numbers": (3,),
+                            },
+                        ],
+                        [
+                            {
+                                "importer": "mypackage.one.alpha.circle",
+                                "imported": "mypackage.green.beta.sphere",
+                                "line_numbers": (8,),
+                            },
+                        ],
+                    ],
+                },
+                {
+                    "upstream_module": "mypackage.green",
+                    "downstream_module": "mypackage.one.alpha.circle",
+                    "chains": [
+                        [
+                            {
+                                "importer": "mypackage.one.alpha.circle",
```

Comment:
Hmm, looking more closely at this test it reveals that the behaviour isn't quite right - it's reporting on the same illegal import twice.

This is because that forbidden modules check the descendants of a given module too. That means that we now have two ways to specify sets of modules - implicitly as descendants and explicitly through wildcards.

Here's an extreme example. Given this contract:

```
[importlinter:contract:forbidden]
name=Forbidden
type=forbidden
source_modules=
    mypackage.blue
    mypackage.blue.*
    mypackage.blue.**
forbidden_modules=mypackage.green
```

The user might get this message:

```
mypackage.blue is not allowed to import mypackage.green:

-   mypackage.blue.alpha.one -> mypackage.green (l.1)


mypackage.blue.alpha is not allowed to import mypackage.green:

-   mypackage.blue.alpha.one -> mypackage.green (l.1)


mypackage.blue.alpha.one is not allowed to import mypackage.green:

-   mypackage.blue.alpha.one -> mypackage.green (l.1)
```

That's three errors but only one illegal import. I think that could be confusing.

Perhaps a better approach would be to exclude any modules matched by wildcards if they are already implicitly being checked by being a subpackage of another module. So in the case above, we'd just see the first error:

```
mypackage.blue is not allowed to import mypackage.green:

-   mypackage.blue.alpha.one -> mypackage.green (l.1)
```

Some extra context - there is an [in-progress PR](https://github.com/seddonym/import-linter/pull/229) to make the behaviour of checking subpackages configurable, so there is some potential for conflicts with that PR.

I don't think this is necessarily a blocker for merge, and _possibly_ not even a blocker for the next release, but wanted to capture my thoughts here. Interested to know what you think.

Reply:
Ah, interesting, I did not consider this behavior.
But is this something introduced by these changes? 

If a user instead specified this
```
[importlinter:contract:forbidden]
name=Forbidden
type=forbidden
source_modules=
    mypackage.blue
    mypackage.blue.alpha
    mypackage.blue.alpha.one
forbidden_modules=mypackage.green
```

Wouldn't the result by the same (i.e. the same illegal import showing thrice)?

> Perhaps a better approach would be to exclude any modules matched by wildcards if they are already implicitly being checked by being a subpackage of another module.

So instead of checking for wildcards in the input, we could just check that the resulting illegal imports in the output are unique?

Reply:
> Wouldn't the result by the same (i.e. the same illegal import showing thrice)?

I just tried it on `master` and you are right, it does! I suppose it's less likely though when people are specifying things explicitly.

I'm happy for us not to worry about this for now.